File: java/BuilderPattern.java
package pattern.java;

/*
 * 
 * Builder Pattern is used so that we can re-use same "construction" code
 * to create "similar" type of objects ( or, Objects of same family).
 * 
 */


abstract class ObjectBuilder{
	private String id;
	private String header;
	private String body;
	private String footer;
	private String type;
	
	public String getType() {
		return type;
	}
	public void setType(String type) {
		this.type = type;
	}
	
	public String getId() {
		return id;
	}
	public void setId(String id) {
		this.id = id;
	}
	public String getHeader() {
		return header;
	}
	public void setHeader(String header) {
		this.header = header;
	}
	public String getBody() {
		return body;
	}
	public void setBody(String body) {
		this.body = body;
	}
	public String getFooter() {
		return footer;
	}
	public void setFooter(String footer) {
		this.footer = footer;
	}
	void buildId(String Id){
		this.setId(Id);
	}
	void buildHeader(String headerText){
		this.setHeader(headerText);
	}
	void buildBody(String bodyText){
		this.setBody(bodyText);
	}
	void buildFooter(String footerText){
		this.setFooter(footerText);
	}
	void printPageData(){
		System.out.println("Details : " + this.getType() + "|header=" + this.getHeader());
	}
}

class ObjectDirector{
	//it needs a construct() with 
	// an Object builder, a page type parameter, and 
	//[optional second argument/arguments 
	//with parameters] to set various page parameters
	public void construct(ObjectBuilder builder,String typeOfPage, String... parameters){
		buildObjectData(builder,parameters.length,parameters);
		builder.setType(typeOfPage);
		builder.printPageData();
	}
	
	public void buildObjectData(ObjectBuilder builder,int length,String [] parameters){
		if(length==1){
			builder.buildId(parameters[0]);
		}else if(length==2){
			builder.buildId(parameters[0]);
			builder.buildHeader(parameters[1]);
		}else if(length==3){
			builder.buildId(parameters[0]);
			builder.buildHeader(parameters[1]);
			builder.buildFooter(parameters[2]);
		}else if(length==4){
			builder.buildId(parameters[0]);
			builder.buildHeader(parameters[1]);
			builder.buildBody(parameters[2]);
			builder.buildFooter(parameters[3]);
		}else{
			return;
		}
	}
}

class HomePage extends ObjectBuilder{
	HomePage(){
		System.out.println("This is home...init ");
	}
}
class LandingPage extends ObjectBuilder{
	LandingPage(){
		System.out.println("This is landing...init");
	}
}

public class BuilderPattern{
	public static void main(String args[]){
		ObjectDirector director = new ObjectDirector();
		director.construct(new HomePage(), "HomePage", "Home Page","This is Home Page Body","This is Home Page Footer");
		director.construct(new LandingPage(), "LandingPage", "Landing Page","This is Landing Page Body","This is Landing Page Footer");
	}
}

File: go/BuilderPattern.go
 package main

import(
	"fmt"
	)
type Page interface{
	getId() string
	setId(id string) 
	getHeader() string
	setHeader(header string)
	getBody() string
	setBody(body string)
	setFooter(footer string)
	getObjType() string
	setType(objType string)
	printPageData()
}
type ObjectBuilder struct{
	id string
	header string
	body string
	footer string
	objType string
}
func(O ObjectBuilder) getId() string{
	return O.id
}
//pass a pointer to struct so that we can modify the data in memory location
func(O *ObjectBuilder) setId(id string){
	O.id = id
}
func(O ObjectBuilder) getHeader() string{
	return O.header
}
func(O *ObjectBuilder) setHeader(header string){
	O.header = header
}
func(O ObjectBuilder) getBody() string{
	return O.body
}
func(O *ObjectBuilder) setBody(body string){
	O.body = body
}
func(O ObjectBuilder) getFooter() string{
	return O.footer
}
func(O *ObjectBuilder) setFooter(footer string){
	O.footer = footer
}
func(O ObjectBuilder) getObjType() string{
	return O.objType
}
func(O *ObjectBuilder) setType(objType string){
	O.objType = objType
}
func(O ObjectBuilder) printPageData(){
	outputString := "From ObjectBuilder##id=" 
	outputString += O.getId()
	outputString += "|Header##"
	outputString += O.getHeader()
	outputString += "|Body##"
	outputString += O.getBody()
	fmt.Println(outputString)
}
type ObjectDirector struct{}
func(O ObjectDirector) construct(builder Page,objType string,parameters ...string){
	builder.setType(objType)
	O.buildObjectData(builder,len(parameters),parameters)
	builder.printPageData()
}
func(O ObjectDirector) buildObjectData(builder Page,length int,parameters []string){
	fmt.Println(parameters)
	if length==1 {
		builder.setId(parameters[0])
	}else if length==2 { 
		fmt.Println("inside length 2")
		i := 0
		for _,param := range(parameters) {
		  fmt.Println("for loop...")
		  if i==0 { 
		     builder.setId(param)
	      } else {
	     	 builder.setHeader(param)
		  }
	      i++
 	   }
	} else if length==3 {
		i := 0
		for _,param := range(parameters) {
		     if i==0 {  
		     	builder.setId(param)
		   	} else if i==1 { 
		     	builder.setHeader(param)
		   	} else if i==2 {
		     	builder.setBody(param)
		   	}
	     i++
       }
   }	
	}//buildObjectData
type HomePage struct{
	ObjectBuilder
} 
type LandingPage struct{
	ObjectBuilder
}
type BuilderPattern struct{}

func main(){
	director := ObjectDirector{}
	director.construct(new(HomePage),"HomePage","1","HomePage Header","HomePage Body")
	director.construct(new(LandingPage),"LandingPage","2","LandingPage Header")
}